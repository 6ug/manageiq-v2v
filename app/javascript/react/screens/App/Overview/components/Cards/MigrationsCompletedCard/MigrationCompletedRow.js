import moment from 'moment';
import React from 'react';
import PropTypes from 'prop-types';
import {
  Grid,
  Icon,
  ListView,
  OverlayTrigger,
  Tooltip
} from 'patternfly-react';

const MigrationCompletedRow = ({ migration }) => {
  const successfulMigrations = migration.miq_request_tasks.filter(
    task => task.state === 'finished'
  ).length;
  const failedMigrations = migration.miq_request_tasks.filter(
    task => task.state !== 'finished'
  ).length;

  const startTime = new Date(migration.options.delivered_on);
  const endTime = Date.now();
  let hours = moment(endTime).diff(startTime, 'hours');
  const minutes = moment(endTime).diff(startTime, 'minutes') % 60;
  const days = Math.floor(hours / 24);
  hours %= 24;

  let elapsedTime;
  if (days >= 2) {
    elapsedTime = sprintf(__('%s days %s:%s'), days, hours, minutes);
  } else if (days === 1) {
    elapsedTime = sprintf(__('1 day %s:%s'), hours, minutes);
  } else {
    elapsedTime = sprintf(__('%s:%s'), hours, minutes);
  }

  return (
    <div className="list-group-item">
      <Grid.Row className="completed-migration-item">
        <Grid.Col xs={8} md={3} className="completed-migration-name">
          <OverlayTrigger
            overlay={
              <Tooltip id={`description_${migration.id}`}>
                {migration.description}
              </Tooltip>
            }
            placement="top"
            trigger={['hover']}
            delay={500}
            rootClose={false}
          >
            <strong>{migration.description}</strong>
          </OverlayTrigger>
        </Grid.Col>
        <Grid.Col xs={4} md={3} mdPush={6}>
          <ListView.InfoItem>
            <Icon type="fa" name="clock-o" />
            <span>{elapsedTime}</span>
          </ListView.InfoItem>
        </Grid.Col>
        <Grid.Col xs={8} md={3} mdPull={3}>
          <ListView.InfoItem>
            <Icon
              type="pf"
              name="ok"
              className={successfulMigrations === 0 ? 'invisible' : ''}
            />
            <strong>{successfulMigrations}</strong>
            {__('Successful')}
          </ListView.InfoItem>
        </Grid.Col>
        <Grid.Col xs={4} md={3} mdPull={3}>
          <ListView.InfoItem>
            <Icon
              type="pf"
              name="error-circle-o"
              className={failedMigrations === 0 ? 'invisible' : ''}
            />
            <strong>{failedMigrations}</strong>
            {__('Failed')}
          </ListView.InfoItem>
        </Grid.Col>
      </Grid.Row>
    </div>
  );
};

MigrationCompletedRow.propTypes = {
  migration: PropTypes.object
};
MigrationCompletedRow.defaultProps = {
  migration: {}
};

export default MigrationCompletedRow;
