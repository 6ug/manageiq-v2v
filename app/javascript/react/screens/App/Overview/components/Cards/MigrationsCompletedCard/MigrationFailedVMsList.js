import React from 'react';
import PropTypes from 'prop-types';

import { Grid, OverlayTrigger, Tooltip } from 'patternfly-react';

const MigrationFailedVMsList = ({ migration }) => {
  const failedMigrations = migration.miq_request_tasks.filter(
    task => task.status === 'Error'
  );

  const getVMNameFromDescription = description => {
    const start = description.indexOf('[');
    const end = description.indexOf(']');

    if (start >= 0 && start < end) {
      return description.slice(start + 1, end);
    }

    return description;
  };

  return (
    <React.Fragment>
      <Grid fluid>
        <Grid.Row>
          <Grid.Col xs={4} md={3}>
            <strong>{__('VM Name')}</strong>
          </Grid.Col>
          <Grid.Col xs={8} md={9}>
            <strong>{__('Error Message')}</strong>
          </Grid.Col>
        </Grid.Row>
      </Grid>
      <Grid fluid className="failures-list">
        {failedMigrations.map(failure => (
          <Grid.Row key={failure.id}>
            <OverlayTrigger
              overlay={
                <Tooltip id={`failed-vm-name-${failure.id}`}>
                  {getVMNameFromDescription(failure.description)}
                </Tooltip>
              }
              placement="top"
              trigger={['hover']}
              delay={500}
              rootClose={false}
            >
              <Grid.Col className="failed-vm-column-text" xs={4} md={3}>
                {getVMNameFromDescription(failure.description)}
              </Grid.Col>
            </OverlayTrigger>
            <OverlayTrigger
              overlay={
                <Tooltip id={`failed-vm-name-${failure.id}`}>
                  {failure.message}
                </Tooltip>
              }
              placement="top"
              trigger={['hover']}
              delay={500}
              rootClose={false}
            >
              <Grid.Col className="failed-vm-column-text" xs={8} md={9}>
                {failure.message}
              </Grid.Col>
            </OverlayTrigger>
          </Grid.Row>
        ))}
      </Grid>
    </React.Fragment>
  );
};

MigrationFailedVMsList.propTypes = {
  migration: PropTypes.object
};
MigrationFailedVMsList.defaultProps = {
  migration: {}
};

export default MigrationFailedVMsList;
