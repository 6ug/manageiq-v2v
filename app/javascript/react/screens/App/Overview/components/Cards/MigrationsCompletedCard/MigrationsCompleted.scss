.completed-migrations-list {
  max-height: 200px;
  overflow-y: auto;

  .list-group-item {
    padding: 0;
  }
}
.completed-migration-item {
  margin: 10px -20px;
  width: 100%;
  .list-view-pf-additional-info-item {
    text-align: left;
    line-height: 26px;
  }
}
.completed-migration-name {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  @media(min-width: 992px) {
    margin-top: 4px;
  }
}
