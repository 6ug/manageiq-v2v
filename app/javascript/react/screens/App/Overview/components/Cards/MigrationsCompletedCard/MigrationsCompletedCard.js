import React from 'react';

const MigrationsCompletedCard = () => (
  <div className="card-pf">
    <div className="card-pf-heading">
      <h2>0 Completed Migrations</h2>
    </div>
    <div className="card-pf-body">
      <p>content</p>
    </div>
  </div>
);

MigrationsCompletedCard.propTypes = {};

export default MigrationsCompletedCard;
