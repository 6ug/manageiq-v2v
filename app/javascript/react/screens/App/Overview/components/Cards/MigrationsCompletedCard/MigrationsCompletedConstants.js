export const FETCH_V2V_MIGRATIONS_COMPLETED = 'FETCH_V2V_MIGRATIONS_COMPLETED';
