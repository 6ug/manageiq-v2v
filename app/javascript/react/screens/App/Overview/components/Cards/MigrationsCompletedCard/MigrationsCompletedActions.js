import URI from 'urijs';
import API, { globalMockMode } from '../../../../../../../common/API';

import { FETCH_V2V_MIGRATIONS_COMPLETED } from './MigrationsCompletedConstants';
import { requestCompletedServiceRequests } from './migrationsCompleted.fixtures';

const mockMode = globalMockMode;

const _getMigrationsCompletedAction = url => dispatch => {
  if (mockMode) {
    dispatch({
      type: `${FETCH_V2V_MIGRATIONS_COMPLETED}_FULFILLED`,
      payload: requestCompletedServiceRequests.response
    });
    return;
  }
  dispatch({
    type: FETCH_V2V_MIGRATIONS_COMPLETED,
    payload: API.get(url)
  });
};

const fetchMigrationsCompletedAction = () => {
  const url =
    '/api/service_requests?' +
    "filter[]=state='finished'" +
    "&filter[]=type='ServiceTemplateTransformationPlanRequest'" +
    '&expand=resources' +
    '&attributes=status,state,options,description,miq_request_tasks' +
    '&sort_by=updated_on';

  const uri = new URI(url);

  return _getMigrationsCompletedAction(uri.toString());
};

export { fetchMigrationsCompletedAction };
