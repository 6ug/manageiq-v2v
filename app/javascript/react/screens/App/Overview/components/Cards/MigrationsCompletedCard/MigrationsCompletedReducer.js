import Immutable from 'seamless-immutable';

import { FETCH_V2V_MIGRATIONS_COMPLETED } from './MigrationsCompletedConstants';

const initialState = Immutable({
  migrationsCompleted: [],
  isFetchingMigrationsCompleted: false,
  isRejectedMigrationsCompleted: false,
  errorMigrationsCompleted: null
});

export default (state = initialState, action) => {
  switch (action.type) {
    case `${FETCH_V2V_MIGRATIONS_COMPLETED}_PENDING`:
      return state.set('isFetchingMigrationsCompleted', true);
    case `${FETCH_V2V_MIGRATIONS_COMPLETED}_FULFILLED`:
      return state
        .set('migrationsCompleted', action.payload.data.resources)
        .set('isFetchingMigrationsCompleted', false);
    case `${FETCH_V2V_MIGRATIONS_COMPLETED}_REJECTED`:
      return state
        .set('errorMigrationsCompleted', action.payload)
        .set('isRejectedMigrationsCompleted', true)
        .set('isFetchingMigrationsCompleted', false);
    default:
      return state;
  }
};
