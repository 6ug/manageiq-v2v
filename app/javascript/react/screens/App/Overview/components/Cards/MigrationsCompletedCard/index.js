import { connect } from 'react-redux';

import MigrationsCompletedCard from './MigrationsCompletedCard';
import reducer from './MigrationsCompletedReducer';
import * as MigrationsCompletedActions from './MigrationsCompletedActions';

export const reducers = { migrationsCompleted: reducer };

const mapStateToProps = ({ migrationsCompleted }, ownProps) => ({
  ...migrationsCompleted,
  ...ownProps.data
});

const mergeProps = (stateProps, dispatchProps, ownProps) =>
  Object.assign(stateProps, ownProps.data, dispatchProps);

export default connect(mapStateToProps, MigrationsCompletedActions, mergeProps)(
  MigrationsCompletedCard
);
