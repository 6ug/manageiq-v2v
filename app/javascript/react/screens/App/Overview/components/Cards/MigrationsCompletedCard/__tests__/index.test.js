import React from 'react';
import thunk from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { initialState } from '../migrationsCompleted.fixtures';
import MigrationsCompletedCard from '../index';
import reducer from '../MigrationsCompletedReducer';

const reducers = { migrationsCompleted: reducer };

describe('MigrationsCompletedCard integration test', () => {
  const middlewares = [thunk, promiseMiddleware()];
  const generateStore = () =>
    createStore(
      combineReducers({ ...reducers }),
      {
        migrationsCompleted: initialState
      },
      applyMiddleware(...middlewares)
    );

  const mountComponent = store =>
    mount(
      <Provider store={store}>
        <MigrationsCompletedCard />
      </Provider>
    );

  it('should mount the MigrationsCompletedCard with mapStateToProps reduced', () => {
    const store = generateStore();
    const wrapper = mountComponent(store);

    // query the unconnected component to assert reduced props
    const component = wrapper.find(MigrationsCompletedCard);
    expect(component.props()).toMatchSnapshot();
  });
});
