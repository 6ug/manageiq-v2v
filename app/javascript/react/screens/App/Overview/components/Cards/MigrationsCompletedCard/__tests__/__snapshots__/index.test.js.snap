// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MigrationsCompletedCard integration test should mount the MigrationsCompletedCard with mapStateToProps reduced 1`] = `Object {}`;
